KIPs/kip-160.md

```diff
@@ -0,0 +1,72 @@
+
+## Specification
+Before proceeding, this proposal will first organize and clarify the terminology. The previous/new fund addresses is called here as `Zeroed`, `Allocated` unlike in Kip103. `Retired`, `Newbie` was ambiguous enough to degrade the readability as they didn't clearly indicate that the previous fund balance is zeroed and the new fund balance is allocated.
+
+### To consider both total burn/total mint cases
+In kip103, the rebalancing only considers the total burn case where the total balance decreases. To make it generalized, this proposal removes the condition that the total balance of `Zeroeds` should be more than the total minting amount of `Allocateds`. As a result, the checking code is removed from the contract and core code.
+
+In the `TreasuryRebalanceV2` contract, the finalizeApproval has been revised like below. Originally, this require statement, `require(getTreasuryAmount() < sumOfZeroedBalance())`, was there. However, it is removed to support all the cases.
+
+```solidity
+   /**
+    * @dev finalizeApproval sets the status to Approved,
+    *      After this stage, approvals will be restricted.
+    */
+   function finalizeApproval()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Registered)
+    {
+        checkZeroedsApproved();
+        status = Status.Approved;
+        emit StatusChanged(status);
+    }
+```
+
+The next validation has been removed from the core logic also. It ensured the total balance of `Zeroeds` is bigger than the total balance of `Allocateds`, but it is removed to support all the cases.
+* totalZeroedAmount >= totalAllocatedAmount
+
+### To make RebalanceBlocknumber defined in TreasuryRebalance contract editable
+In the treasury rebalance contract, there's a `RebalanceBlocknumber` storage and it should be matched with the related hard fork block number. However, even though the hard fork block number is updatable, the RebalanceBlocknumber field cannot be edited. To ensure the same behavior with the hard fork block number, the field become editable in this proposal.
+
+The next method is added to the TreasuryRebalanceV2 contract.
+```solidity
+    /**
+     * @dev updates rebalance block number
+     * @param _rebalanceBlockNumber is the updated target block number of the execution the rebalance in Core
+     */
+    function updateRebalanceBlocknumber(
+        uint256 _rebalanceBlockNumber
+    ) public onlyOwner {
+        require(
+            block.number < rebalanceBlockNumber && 
+            rebalanceBlockNumber < _rebalanceBlockNumber
+        );
```

Comment:
Why do we only support higher `_rebalanceBlockNumber`? To cover all cases, I think `block.number < rebalanceBlockNumber && block.number < _rebalanceBlockNumber && rebalanceBlockNumber != _rebalanceBlockNumber` would be enough.

Reply:
It's because normally we update the hard fork block number to higher number, but your comment makes sense. I'll update it.

@hyeonLewis What about this? checking two conditions respectively and return with error.
```
        require(block.number < rebalanceBlockNumber, "current block shouldn't be past the currently set block number")
        require(block.number < _rebalanceBlockNumber, "rebalance blockNumber should be greater than current block");
```

Reply:
Sorry for being late. LGTM!

Reply:
Here's the fix: https://github.com/klaytn/kips/pull/160/commits/2266ce758922cb50cea91d5b95800cf3a7198720